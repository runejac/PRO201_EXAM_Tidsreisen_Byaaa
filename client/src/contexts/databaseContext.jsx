import React, { createContext } from "react";
import { fetchJSON_client } from "../helpers/http.jsx";

export const DatabaseContext = createContext({
  async listHistory() {
    return await fetchJSON_client("/api/history");
  },
  async listQuiz(query) {
    return await fetchJSON_client("/api/quiz?" + new URLSearchParams(query));
  },
  async listAudio() {
    return await fetchJSON_client("/api/sound");
  },
});
