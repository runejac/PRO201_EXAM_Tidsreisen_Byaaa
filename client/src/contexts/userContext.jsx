import React, { createContext } from "react";
import { fetchJSON_client, putJSON } from "../helpers/http.jsx";

export const UserContext = createContext({
  async getUser() {
    return await fetchJSON_client("/api/login");
  },
  async updateUser(user) {
    return await putJSON("/api/login/updateuser", user);
  },
});
