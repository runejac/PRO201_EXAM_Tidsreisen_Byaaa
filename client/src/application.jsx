import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
