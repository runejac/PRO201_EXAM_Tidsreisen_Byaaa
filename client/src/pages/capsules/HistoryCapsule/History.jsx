import React, { useContext, useState } from "react";
import { DatabaseContext } from "../../../contexts/databaseContext.jsx";
import { useLoading } from "../../../helpers/useLoading.jsx";
import { Loading } from "../../../components/Loading.jsx";
import { StoryCard } from "./StoryCard.jsx";
import { NotLoggedIn } from "../../../components/NotLoggedIn.jsx";
import { User } from "../../../application.jsx";
import { ErrorModal } from "../../../components/ErrorModal.jsx";
import sawEffect from "../../../assets/images/saw-effect.png";
import "./storycard.css";
import "../../../css/olaf.css";

export function History() {
  const { user } = useContext(User);
  const { name } = user;

  const [error, setEror] = useState();
  const { listHistory } = useContext(DatabaseContext);
  const {
    data,
    error: errorHistory,
    loading,
  } = useLoading(async () => await listHistory());

  if (loading) {
    return <Loading />;
  }

  if (errorHistory) {
    setEror(errorHistory);
    console.error(error);
    return <ErrorModal error={error} />;
  }

  if (name === undefined) {
    return <NotLoggedIn />;
  }

  return (
    <div
      style={{
        backgroundColor: "#F2F1E8",
        height: "100vh",
      }}
    >
      <img
        id={"image-saw-effect"}
        height={700}
        src={sawEffect}
        alt="bilde av vann sag effekt bakgrunn"
      />
      {data && (
        <div>
          {data?.map((historyCapsule, index) => {
            return (
              <div key={index}>
                <StoryCard
                  loading={loading}
                  error={error}
                  historyCapsule={historyCapsule}
                />
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
}
