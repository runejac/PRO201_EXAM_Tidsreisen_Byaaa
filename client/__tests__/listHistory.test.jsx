/**
 * @jest-environment jsdom
 */

import ReactDOM from "react-dom";
import React from "react";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { DatabaseContext } from "../src/contexts/databaseContext.jsx";
import { User } from "../src/application.jsx";
import { UserContext } from "../src/contexts/userContext.jsx";
import { History } from "../src/pages/capsules/HistoryCapsule/History.jsx";

describe("ListHistory", function () {
  it("should show history", async function () {
    const user = {
      name: "Hei",
      points: 23,
      finishedCapsules: [],
      intro: true,
      walk: false,
    };

    const context = jest.fn();

    const history = [
      {
        category: "Historiekapsel",
        story: [
          {
            year: "År 0000",
            story: "test-historie",
            image: "test-bilde",
            done: false,
          },
          {
            year: "År 0000",
            story: "test-historie",
            image: "test-bilde",
            done: false,
          },
          {
            year: "År 0000",
            story: "test-historie",
            image: "test-bilde",
            done: false,
          },
          {
            done: true,
          },
        ],
        name: "testName",
        id: 1,
      },
    ];
    const container = document.createElement("div");

    await act(async () => {
      ReactDOM.render(
        <MemoryRouter>
          <DatabaseContext.Provider value={{ listHistory: () => history }}>
            <User.Provider value={{ user }}>
              <UserContext.Provider value={{ context }}>
                <History />
              </UserContext.Provider>
            </User.Provider>
          </DatabaseContext.Provider>
        </MemoryRouter>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });
});
