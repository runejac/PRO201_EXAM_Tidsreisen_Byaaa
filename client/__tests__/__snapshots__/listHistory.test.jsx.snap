// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListHistory should show history 1`] = `
<div>
  <div
    style="background-color: rgb(242, 241, 232); height: 100vh;"
  >
    <img
      alt="bilde av vann sag effekt bakgrunn"
      height="700"
      id="image-saw-effect"
      src="test-file-stub"
    />
    <div>
      <div />
    </div>
  </div>
</div>
`;
